design/view-change.md

```diff
@@ -0,0 +1,229 @@
+
+# PBFT based view change
+
+## Basic requirements and assumptions
+
+View change should not lose any request that was possibly ordered
+by any node, and client possibly got a reply.
+
+After view change all nodes should be in eventually consistent state.
+
+Primaries are selected in round robin fashion, so primary for any given
+viewNo could be found just from number of nodes and that viewNo.
+
+It is assumed that any message sent will eventually be delivered, although
+we don't place any upper limit on this time.
+
+## Notations
+
+- _i_, _j_ - node ID
+- _v_ - viewNo
+- _k_ - ppSeqNo of batch
+- _d_ - digest of batch
+- _h_ - ppSeqNo of stable checkpoint
```

Comment:
Checkpoints are redunadant and should be removed, the last ordered request can serve as checkpoint.
PBFT requires checkpoints as a mechanism that is triggered periodically to make sure every node has processed (and persisted) the same requests (across batches) in the same order.
We do not require checkpoints since Pre-Prepare contains expected merkle roots (txn and state merkle tr(i/e)e roots after applying the requests of the PrePrepare) and a Pre-Prepare can be only processed if the previous Pre-Prepare was received and processed successfully (not committed necessarily).
Thus it cannot happen that a node skips processing any PrePrepare.
To handle the case where a node gets a Pre-Prepare(s) and maybe Prepares too but is not getting sufficient Commits, a node can use `MESSAGE_REQUEST` to request missing 3 phase messages or start catchup once it has seen `n` Pre-Prepares (and maybe some Prepares too) but not been able to order them or it has >2f COMMIT(s) for a `pp_seq_no` that is greater than its last ordered sequence number by `n`

Reply:
I'm not sure that this is more efficient and easier than Checkpoints, especially when the network is not stable and we have quite a lot of missing requests.

Reply:
I believe that main purpose of checkpoints is to enable garbage collection of message log, which in turn is used for out-of-order request processing and for message resending, so I don't see how checkpoints can be dropped.